csrc/device_lower/validation.cpp

```diff
@@ -206,6 +206,191 @@ void validateCpAsyncBulk(const std::vector<TensorView*>& tvs) {
+      frontier.erase(inner_it);
+
+      // If it's contig, we can continue the analysis by proceeding to
+      // the output. If not, no further analysis is possible, so the
+      // two inputs are just removed from the frontier list
+      if (is_contig) {
+        frontier[outer_pos] = merge->out();
+      } else {
+        frontier.erase(outer_it);
+      }
+    } else if (auto split = dynamic_cast<Split*>(expr)) {
+      auto in_it = std::ranges::find(frontier, split->in());
+      if (in_it == frontier.end()) {
+        continue;
+      }
+      frontier.insert(in_it + 1, split->inner());
+      *in_it = split->outer();
+    }
+  }
+
+  // Once the traversal is done, if the target id located at the
+  // rightmost position of the frontier list, it is guaranteed to
+  // correspond to the innermost subregion of the base ID.
+  return !frontier.empty() && frontier.back() == maybe_innermost_id;
+}
+
+// Expr-specific validaion
+//
+// TODO: Move individual validations to here, e.g.,
+// validateCpAsyncBulk can be moved here
+class ExprValidator : public OptOutDispatch {
+ public:
+  static void validate(Fusion* fusion) {
+    ExprValidator validator(fusion);
+  }
+
+ private:
+  ExprValidator(Fusion* fusion) {
+    for (auto expr : fusion->exprs()) {
+      dispatch(expr);
+    }
+  }
+
+  // The lowering of ArgsortOp depends on specific scheduling
+  // assumptions. This tight coupling isn't ideal, but it's not a
+  // problem for now.
+  void handle(ArgsortOp* aop) final {
+    auto inp_tv = ir_utils::getTvInput(aop);
+    auto out_tv = ir_utils::getTvOutput(aop);
+
+    // Both input and output must be Local tensors. The local input
+    // tensor is initialized to the max or min value so that
+    // out-of-bounds items do not affect the results
+    NVF_ERROR_EQ(inp_tv->getMemoryType(), MemoryType::Local);
+    NVF_ERROR_EQ(out_tv->getMemoryType(), MemoryType::Local);
+
+    // The input will be initialized for this argsort. To avoid any
+    // potential conflict of initialization, require the input to be
+    // exclusively used by the argsort
+    NVF_ERROR_EQ(inp_tv->uses().size(), 1);
+
+    IterDomain* grouped_id = nullptr;
+    for (const auto& loop_id : out_tv->getLoopDomain()) {
+      if (loop_id->getParallelType() == ParallelType::Group) {
+        NVF_ERROR(grouped_id == nullptr, "Multiple IDs found to be grouped");
+        grouped_id = loop_id;
+      }
+    }
+
+    if (grouped_id != nullptr) {
+      // If it's grouped, it must correspond to the innermost subregion
+      // of the logical argsort ID
+      IterDomain* argsort_logical_id =
+          out_tv->getLogicalDomain().at(aop->dim());
+      NVF_ERROR(
+          isInnermost(argsort_logical_id, grouped_id),
+          "Invalid ID to group: ",
+          grouped_id->toString(),
+          " of ",
+          aop->toString());
+
+      // The output is allocated on registers, so the allocation
+      // domain should be just the same as the loop domain. The
```

Comment:
> so the allocation domain should be just the same as the loop domain.

IIUC, we are discarding the allocation domain for register TVs. So we don't really require the allocation domain to be the same and it's safe to only check the loop domain.